### ISSUE TITLE
Issue with JavaScript toolkit since version 2.3.0

### ISSUE THREAD

**Issue by WolfgangDrescher** [rism-ch/verovio#1199]
I have a setup where verovio is loaded a npm dependency and provided as a webpack externals. Trying to instance an new toolkit with `let vrvToolkit = new verovio.toolkit();` i get the following error message:

`TypeError: Cannot read property 'apply' of undefined at Object.Module._vrvToolkit_constructor [as constructor] (index.js:95)`
 
It seems like the problem is that the variable `Module["asm"]["B"]` in undefined.

Downgrading to version 2.2.1 solved this issue for me.

**Comment by lpugin** [comment]
Could you try to do ?
```
var verovio = require("verovio").init()
```

**Comment by WolfgangDrescher** [comment]
First i got a error message `init is not a function` but probably because i was using verovio with the CopyWebpackPlugin and as a webpack external. Therefore JavaScript was trying to interpret the init function and not Node.js. But changing this in my config, webpack could handle the huge verovio file and i got the same error (with version 2.3.1).

Hotlinking `<script src="http://www.verovio.org/javascript/latest/verovio-toolkit.js">` (`2.3.2-50dad07`) worked fine. So maybe its a problem with CopyWebpackPlugin or my setup.

**Comment by lpugin** [comment]
OK, thanks for the update. The problem seems to come from this option that allows for the memory to be set through the init function. It is actually a bit hacky and it would be best to remove it. I am not sure how many people make use of it. Some users will probably complain if we remove it but they can still compile their own module if necessary.

**Comment by kepper** [comment]
since I'm the one who will complain 😉 – I'm running into problems with this mechanism anyway, as my Vue setup always tries to re-compile the Verovio code, which just takes a lot of time, but doesn't result in better / more condensed code… If that mechanism is hacky, at this point I don't mind removing it, given that there are good instructions on how to compile a version with more RAM assigned, ideally in a Docker and / or NodeJS environment…

**Comment by lpugin** [comment]
@WolfgangDrescher please check with 2.3.3

**Comment by WolfgangDrescher** [comment]
I downloaded the toolkit manually (`2.3.3-654119c`) and on the first sight it seems to work now. But i will check it again as soon as you have released the new version on npm. Thanks for the update!

In 2.2.1 i frequently got errors where module is not defined so i tought that maybe it would be a good idea to wrap verovio in a factory function like for example jQuery does it:

```js
(function(global, factory) {
	if (typeof module === "object" && typeof module.exports === "object") {
		// For CommonJS and CommonJS-like environments where a proper `window`
		// is present, execute the factory and get verovio.
		// For environments that do not have a `window` with a `document`
		// (such as Node.js), expose a factory as module.exports.
		// This accentuates the need for the creation of a real `window`.
		// e.g. var verovio = require("verovio")(window);
		module.exports = global.document ?
			factory(global, true) :
			function(w) {
				if(!w.document) {
					throw new Error("Verovio requires a window with a document");
				}
				return factory(w);
			};
	} else {
		factory(global);
	}

// Pass this if window is not defined yet
} )(typeof window !== "undefined" ? window : this, function(window, noGlobal) {
	// verovio source ...
	return verovio;
});
```

But version 2.3 seems to solve this problem anyways.

**Comment by lpugin** [comment]
OK. Re-open if necessary

**Comment by WolfgangDrescher** [comment]
@kepper Im also using verovio in a Vue setup but im not so happy with mine so i would be interested to know how your setup looks like.

As you wrote re-compiling with hot reload take a lot of time and a few months ago my webpack setup could not handle the 10MB heavy verovio file (seems to work now). Thats why im just copying the current version of verovio (loaded as a npm dependency) with the CopyWebpackPlugin and then import it as a webpack external variable.

So simplified my setup looks like this:

```js
// vue.config.js
const CopyWebpackPlugin = require('copy-webpack-plugin');

module.exports = {
    configureWebpack: {
        externals: {
            verovio: 'verovio',
        },
        plugins: [
            new CopyWebpackPlugin([{
                from: 'node_modules/verovio',
                to: 'verovio',
            }]),
        ],
    },
};
```

```html
<!-- head in index.html -->
<script src="<%= BASE_URL %>verovio/index.js"></script>
```

```vue
<!-- Score.vue -->
<template>
    <div v-html="renderedScore"></div>
</template>

<script>
import verovio from 'verovio'; // load verovio as a webpack external

export default {
    data() {
        return {
            verovioToolkit: null,
            renderedScore: null,
            scale: 40,
            page: 1,
        };
    },
    created() {
        this.verovioToolkit = new verovio.toolkit();
    },
    mounted() {
        this.setVerovioOptions();
        this.loadScoreFile('/path/to/score.mei');
    },
    watch: {
        scale() {
            this.rerenderCurrentView();
        },
        page(value) {
            this.rerenderCurrentView();
        },
    },
    methods: {
        setVerovioOptions() {
            this.verovioToolkit.setOptions({
                // adjustPageHeight: true,
                scale: this.scale,
                noHeader: true,
                noFooter: true,
                pageWidth: this.width * (100 / this.scale),
                pageHeight: this.height * (100 / this.scale),
            });
        },
        async loadScoreFile(path) {
            const response = await fetch(path);
            const data = await response.text();
            this.renderedScore = this.verovioToolkit.renderData(data, {});
        },
        rerenderCurrentView() {
            this.setVerovioOptions();
            this.verovioToolkit.redoLayout();
            this.renderedScore = this.verovioToolkit.renderToSVG(this.page, {});
        },
    },
};
</script>

**Comment by pfefferniels** [comment]
Unfortunately I'm still getting this `TypeError: Cannot read property 'apply' of undefined
    at Object.Module._vrvToolkit_constructor [as constructor] (/index.js:73:100412)` after updating to version 2.3.3, with 2.2.0 everything worked fine. I suppose it is related to this issue ...

**Comment by lpugin** [comment]
See https://github.com/rism-ch/verovio/issues/1247#issuecomment-570592845

**Comment by WolfgangDrescher** [comment]
@lpugin Im having the same issue again in`2.4.0`. I just checked for the most currenct versions:
`2.4.0`, `2.3.5`, `2.3.4-1` all seem not to work while `2.3.3-1` and `2.3.3` are working just fine.

I also tried it with the new method documented on npm: `verovio.module.onRuntimeInitialized = function () {…}`. Im getting `Cannot set property 'onRuntimeInitialized' of undefined`. Is this way also correct to use when not running a node application but a JavaScript App? For me it looks like the verovio variable im getting with `import verovio from "verovio"` does not have a property `module`.

**Comment by WolfgangDrescher** [comment]
I did more research and maybe it has something to do with my setup with verovio as a webpack external variable. Nonetheless it look like currently it is not possible to use verovio with vanilla JavaScript since node needs to be running to detect when the module is ready and `onRuntimeInitialized` is emitted.

**Comment by lpugin** [comment]
`verovio.module` is not available before 2.3.5. Waiting for `onRuntimeInitialized` is the way to load it. This should work (and works with https://npm.runkit.com/verovio)
```
var verovio = require( 'verovio' );

/* Wait for verovio to load */
verovio.module.onRuntimeInitialized = function ()
{
    /* create the toolkit instance */
    var vrvToolkit = new verovio.toolkit();
    console.log(vrvToolkit.getVersion());
}
```

**Comment by WolfgangDrescher** [comment]
Thank for your fast support. Yes i tried this, but vanilla JavaScript will not export the module since `module.exports` is a Node.js function as far as i understand it. So if I just use verovio with a default script tag `<script src="path/to/verovio/from/npm/index.js"></script>` there will not be a module exported. And the `onRuntimeInitialized` will never be called.

Because webpack can currently not handle verovio well as a Node.js dependency since it is really big i currently dont see another way of using it other than as a webpack external variable.

One last notice: using a timeout before instantiating a new toolkit is working. So indeed in seems like verovio is just not ready to make an instance.

```js
setTimeout(() => {
    const verovioToolkit = new verovio.toolkit();
    // ...
}, 1000);
```

Is there a way to detect if the emscripten is interpreted without the Node module event you mentioned? Could also relate to #1253.

**Comment by jericho-music** [comment]
@WolfgangDrescher, the following Vue component code worked for me with Verovio 2.4:

```
<template>
  <div class="music-sheet">
    <h1>Music Sheet</h1>
  </div>
</template>

<script>
import verovio from 'verovio';
import Vue from 'vue'

export default {
  name: 'music-sheet',

  data () {
    return {
      verovioToolkit: null,
    }
  },

  created() {
    Vue.$log.debug('Creating verovio toolkit...');
    /* Wait for verovio to load */
    verovio.module.onRuntimeInitialized = function () {
      /* create the toolkit instance */
      this.vrvToolkit = new verovio.toolkit();
      Vue.$log.debug("Verovio toolkit version: ", this.vrvToolkit.getVersion());
    }
  }
}
</script>
```

with the additional caveat that 'veriovio.wasm' copied into `public' folder (a quick fix for now). Have to figure out a better way to redirect when it is requested.

Adding this to the thread in case this helps you or anyone else.

**Comment by nccurry** [comment]
@jericho-music How did you get the above code to work? 

The Vue context isn't available inside the onRuntimeInitialized function, so the following does not set the vrvToolkit data property properly.

```
    verovio.module.onRuntimeInitialized = function () {
      /* the 'this' keyword here does not point to the vue instaance */
      this.vrvToolkit = new verovio.toolkit();
      Vue.$log.debug("Verovio toolkit version: ", this.vrvToolkit.getVersion());
    }
```

**Comment by jericho-music** [comment]
Hi Nick,

  Yes, 'this' will not be in context there. I am not an expert in JS and got tripped up by 'this' in other areas :) Vuex is an option, but we decided to go with another approach and that is to make Verovio a Vue plug-in. This allows for us to use Vue.$verovio (the variable we defined) in the entire application. That also fit in with the rest of our prototype design of using some external dependencies as plug-ins. Taking your Verovio initialization code and converting it to a plug-in is trivial.

Second, we just publish Vue.$verovio.getVersion() to the log. This is more for debugging rather than a semantic check. If something goes wrong with Verovio initialization, that statement tends to be missing from the log. Though, after @WolfgangDrescher helped me in resolving the initial issue, we have never run into any problems with Verovio initialization.

Hope that helps.
